#517161 [C#テキスト]2章演習問題

sanada-being opened the pull request:
2章演習問題の回答を作成しました。

sanada-being committed:
#517161 [C#テキスト]2章演習問題

m-hotta-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,34 @@
+
+            //コンソールにSongオブジェクトの内容を表示
+            foreach (var wSong in wSongs) {
+                Console.WriteLine($"{wSong.Title}は{wSong.ArtistName}の曲です。局の長さは{wSong.Length / 60}:{wSong.Length % 60:00}秒です。");
```
**【MUST: 修正してください。】**
```suggestion
                Console.WriteLine($"{wSong.Title}は{wSong.ArtistName}の曲です。曲の長さは{wSong.Length / 60}:{wSong.Length % 60:00}秒です。");
```

m-hotta-being replied:
👍

m-hotta-being commented on Chapter2/Chapter2-1-1/Song.cs:
```diff
@@ -0,0 +1,35 @@
+    /// </summary>
+    internal class Song {
+
+        /// <summary>
+        /// 曲名
+        /// </summary>
+        public string Title { get; set; }
+
+        /// <summary>
+        /// アーティスト名
+        /// </summary>
+        public string ArtistName { get; set; }
+
+        /// <summary>
+        /// 曲の長さ
+        /// </summary>
+        public int Length { get; set; }
```
【FYI: 情報共有。アクション不要です。】
もしこれらのプロパティが後から変更されることがない想定なら、読み取り専用にしてしまったほうが、変更される可能性を考慮せずに済むぶん読みやすくなります。
```suggestion
        /// <summary>
        /// 曲名
        /// </summary>
        public string Title { get; }

        /// <summary>
        /// アーティスト名
        /// </summary>
        public string ArtistName { get; }

        /// <summary>
        /// 曲の長さ
        /// </summary>
        public int Length { get; }
```

m-hotta-being commented on Chapter2/Chapter2-1-1/Song.cs:
```diff
@@ -0,0 +1,35 @@
+        /// コンストラクタ
+        /// </summary>
+        /// <param name="vTitle">曲名</param>
+        /// <param name="vArtistname">アーティスト名</param>
+        /// <param name="vLength">曲の長さ</param>
+        public Song(string vTitle, string vArtistname, int vLength) {
```
【FYI: 情報共有。アクション不要です。】
`vLength` は分でしょうか？秒でしょうか？あるいはミリ秒でしょうか？
そういった情報も盛り込めると使いやすくなりそうです。

teshima-being replied:
すみません。指摘が漏れていました。
`vArtistname`は`vArtistName`ですね。

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,34 @@
+            //Songクラスのインスタンスを追加
+            wSongs.Add(new Song("エイリアンズ", "キリンジ", 376));
+            wSongs.Add(new Song("aoi", "サカナクション", 253));
+            wSongs.Add(new Song("花束", "サンボマスター", 401));
```
wSongsをList<Song>にしたのですね。Songインスタンスの追加が容易になり良いと思います！

※追加課題
今回のList<Song>とは別に、配列wSongArrayを用意し、そこに3つほどSongインスタンスを格納してみてください。
また、配列とListの違いについてもこの機会に調べてみてください。（すでに確認済ならOKです。）

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,34 @@
+
+            //コンソールにSongオブジェクトの内容を表示
+            foreach (var wSong in wSongs) {
+                Console.WriteLine($"{wSong.Title}は{wSong.ArtistName}の曲です。局の長さは{wSong.Length / 60}:{wSong.Length % 60:00}秒です。");
```
※追加課題
曲の長さの出力について、`{wSong.Length / 60}:{wSong.Length % 60:00}`にしてくれていますが、TimeSpan構造体を使って書き直してみてください。
参考：[TimeSpan構造体](https://learn.microsoft.com/ja-jp/dotnet/api/system.timespan?view=net-7.0)
　　　[【C#】秒数をhh:mm:ss変換する](https://qiita.com/Nossa/items/70487b765ec9332e0db0)

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,34 @@
+
+4. 配列に格納されたすべてのSongオブジェクトの内容をコンソールに表示してください。
+演奏時間の表示は「4:16」のような書式にしてください。ただし、演奏時間は必ず60分未満と仮定してよいです。
+        */
```
コメント部分のインデントは揃っていた方が見やすいと思います。
またusingとnamespaceの間を1行空けておきましょう。

teshima-being commented on Chapter2/Chapter2-1-1/Song.cs:
```diff
@@ -0,0 +1,35 @@
+        public Song(string vTitle, string vArtistname, int vLength) {
+            Title = vTitle;
+            ArtistName = vArtistname;
+            Length = vLength;
```
左辺にthisを付けておきましょう。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,50 @@
+namespace Chapter2_1_3 {
+    //売上集計クラス
+    public class SalesCounter {
+        private IEnumerable<Sale> _sales;
```
salesはprivateなフィールド変数なので[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#%E6%8E%A5%E9%A0%AD%E8%BE%9E)に則り、接頭語Fを付けましょう。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,50 @@
+using System.IO;
+namespace Chapter2_1_3 {
+    //売上集計クラス
+    public class SalesCounter {
```
クラスにはドキュメントコメントを記述しておきましょう。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,50 @@
+        /// </summary>
+        /// <returns>商品別の売上を返す</returns>
+        public IDictionary<string, int> GetPerProductCategorySales() {
+            var wDict = new Dictionary<string, int>();
```
変数名`wDict`についてもっと良いものがないか考えてみてください。
`wDict`という名前からはどのような情報をもっているかが分かりにくいです。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,50 @@
+            var wLines = File.ReadAllLines(vFilepath);
+            foreach (var wLine in wLines) {
+                var wItems = wLine.Split(',');
+                var wSale = new Sale {
+                    ShopName = wItems[0],
+                    ProductCategory = wItems[1],
+                    Amount = int.Parse(wItems[2])
```
※追加課題
現在は初期化子{};でSaleインスタンスの各プロパティを設定していますが、Saleクラスにコンストラクタを実装し、コンストラクタを使って書き直してみてください。
コンストラクタを実装することで、Saleクラスのインスタンスを作成するときに、必要なプロパティを強制的に設定させることができるというメリットがあります。

現在のままだと下記のように、例えば店名がないSaleインスタンスを作成することが出来ますが、コンストラクタを適切に実装することで、そういったことを防ぐことが出来ます。
```
                var wSale = new Sale {
                    ProductCategory = wItems[1],
                    Amount = int.Parse(wItems[2])
                };
```

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,34 @@
+            // インスタンスを格納するList
+            List<Song> wSongs = new List<Song>();
+
+            //Songクラスのインスタンスを追加
```
細かいですが、23行目と28行目のコメントについて、`//`のあとの半角スペースが抜けています。

teshima-being commented on Chapter2/Chapter2-1-2/Program2-1-2.cs:
```diff
@@ -0,0 +1,16 @@
+        static void Main(string[] args) {
+            for (int wInch = 1; wInch <= 10; wInch++) {
+                double wMeter = wInch * 0.0254;
+                Console.WriteLine($"{wInch}inch = {wMeter:0.0000}");
```
出力時にメートルについても単位を出力すると良いですね。
このプログラムを実行した人が、インチがどの単位に変換されたのかが分かりやすくなります。

teshima-being reviewed (commented):
特に問題なくよくできていると思います！
細かい指摘をいくつかと、追加で考えてみて欲しいことをコメントしましたので、確認のほどよろしくお願いします。

sanada-being committed:
#517161  [C#テキスト]2章演習問題

・usingとnamespaceの間に改行を追加しました。
・Chapter2/Chapter2-1-1/Program2-1-1.csにてコメント部分のインデントを整えました。
・Chapter2/Chapter2-1-1/Program2-1-1.cs38行目の局→曲に変更しました。
Chapter2/Chapter2-1-1/Program2-1-1.csにて、//の後に半角スペースを挿入しました。
・Chapter2/Chapter2-1-1/Program2-1-1.cs22行目に配列を作成し、Songインスタンスを３つ格納しました。
・Chapter2/Chapter2-1-1/Program2-1-1.cs38行目にて、曲の時間表記をTimeSpan構造体を使って書き換えました。
・Chapter2/Chapter2-1-1/Song.csにて、プロパティを参照する際にthisを追加しました。
・Chapter2/Chapter2-1-3/SalesCounter.csにて、フィールド変数salesに接頭語Fを付けました。
・Chapter2/Chapter2-1-3/SalesCounter.csにて、クラスにドキュメントコメントを追加しました。
・Chapter2/Chapter2-1-3/SalesCounter.cs42行目にて、変数名wDictをwProductCategorySalesに変更しました。
・Chapter2/Chapter2-1-3/SalesCounter.csにて、Saleクラスにコンストラクタを実装し、コンストラクタを使ってインスタンスを作成しました。
・Chapter2/Chapter2-1-2/Program2-1-2.csにて出力時にメートルについても単位が表示されるように変更しました。

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -1,33 +1,42 @@
+            // コンソールにSongオブジェクトの内容を表示
             foreach (var wSong in wSongs) {
-                Console.WriteLine($"{wSong.Title}は{wSong.ArtistName}の曲です。局の長さは{wSong.Length / 60}:{wSong.Length % 60:00}秒です。");
+                TimeSpan wDuaration = new TimeSpan(0, 0, wSong.Length);
+                Console.WriteLine($"{wSong.Title}は{wSong.ArtistName}の曲です。曲の長さは{new TimeSpan(0, 0, wSong.Length).ToString(@"mm\:ss")}秒です");
```
TimeSpan構造体で書き換えられましたね！
出力時に `{new TimeSpan(0, 0, wSong.Length).ToString(@"mm\:ss")}`とするのであれば、38行目で定義している変数は不要になるので、行ごと削除しておきましょう。

sanada-being replied:
38行目を削除しました。

teshima-being commented on Chapter2/Chapter2-1-1/Song.cs:
```diff
@@ -27,9 +27,9 @@ internal class Song {
         /// <param name="vArtistname">アーティスト名</param>
         /// <param name="vLength">曲の長さ</param>
         public Song(string vTitle, string vArtistname, int vLength) {
```
すみません。前回の確認漏れです。
細かいですが、`vArtistname` は `vArtistName` ですね。

sanada-being replied:
名称を修正しました。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <returns>Saleオブジェクトのリスト</returns>
+        private static IEnumerable<Sale> ReadSales(string vFilepath) {
+            var wSales = new List<Sale>();
+            var wLines = File.ReadAllLines(vFilepath);
+            foreach (var wLine in wLines) {
```
※修正は任意
CSVを読み取った結果を変数に一度格納していますが、特に変数にする理由がなければ、下記のように記述することで変数の宣言の必要がなくなります。
```
            foreach (string wLine in File.ReadAllLines(vFilepath)) {
                var wItems = wLine.Split(',');
                var wSale = new Sale(wItems[0], wItems[1], int.Parse(wItems[2]));
                wSales.Add(wSale);
            }
```
`Chapter2-1-3`の`Program2-1-3.cs`も同様に検討できる箇所があります。
その変数は本当に宣言する必要があるかを都度考える癖をつけるとよいと思います。

sanada-being replied:
SalesCounter.cs , Program2-1-3.cs共に一時的な変数を作成せず、foreach文に直接処理を渡すように変更しました。

teshima-being commented on Chapter2/Chapter2-1-2/Program2-1-2.cs:
```diff
@@ -0,0 +1,17 @@
+        */
+
+        static void Main(string[] args) {
+            for (int wInch = 1; wInch <= 10; wInch++) {
+                double wMeter = wInch * 0.0254;
+                Console.WriteLine($"{wInch}inch = {wMeter:0.0000}m");
```
※参考情報
以前AtCoderをやってみたいという話をされていたので、それに関連した話を少し。
今の実装だとプログラムを実行すると常に1～10インチをメートルに変換する変換表が表示されますが、表示する範囲を実行時に指定できるようになるとより使い勝手がよくなります。
下記のようにConsole.ReadLine()を使うことで、ユーザーの入力値を受け取り、それに応じて表示する範囲を変更することができます。
```
        static readonly double C_Ratio = 0.0254;

        static void Main(string[] args) {

            Console.WriteLine("インチとメートルの対応表を表示します。\n対応表の開始値（単位インチ）を入力してください。");
            int wStart = int.Parse(Console.ReadLine());

            Console.WriteLine("対応表の終了値（単位インチ）を入力してください。");
            int wEnd = int.Parse(Console.ReadLine());

            PrintInchToMeterTable(wStart, wEnd);
        }

        static void PrintInchToMeterTable(int vStart, int vEnd) {
            for (int wInch = vStart; wInch <= vEnd; wInch++) {
                double wMeter = wInch * C_Ratio;
                Console.WriteLine($"{wInch}inch = {wMeter:0.0000}m");
            }
        }
```
Console.ReadLine()での入力値の受付はAtCoderのどの問題でも必須ですので、今から色々試してみるとよいと思います。
また、上記プログラムだとint型以外を入力すると例外が発生します。どうすればそれを回避できるか考えてみるのも面白いと思います。

teshima-being reviewed (commented):
あと少しだけ修正をお願いします。

sanada-being committed:
#517161 [C#テキスト]2章演習問題

再レビューの指摘内容（参考情報除く）を修正しました。

teshima-being commented:
レビューOKです。1章お疲れ様でした。
